File: leetcode/generate-parentheses.cpp
#include <string>
#include <stack>
class Solution {
public:
    bool well_formed_parentheses(string a){
        stack<int> s;
        for(int i=0;i<a.length();i++){
            if(s.empty()) s.push(a[i]);
            else if (s.top()=='(' && a[i]==')') s.pop();// continue;
            else s.push(a[i]);
        }
        return s.empty();
    }
    void BackTracking(vector<string> &answer,int n,int curIndex=0,string a=""){
        if(2*n==curIndex) {
            if(well_formed_parentheses(a)) answer.push_back(a);
        }
        else{
            BackTracking(answer,n,curIndex+1,a+'(');
            BackTracking(answer,n,curIndex+1,a+')');
        }
        
    }
    vector<string> generateParenthesis(int n) {
        vector<string> answer;
        BackTracking(answer,n);
        return answer;
    }
};

File: leetcode/generate-parentheses.py
import itertools
class Solution:
    def isPair(self, parentheses):
        stack = []
        for i in parentheses:
            if stack and stack[-1] == '(' and i == ')':
                stack.pop()
            else:
                stack.append(i)
        return True if not stack else False
    
    def generateParenthesis(self, n: int) -> List[str]:
        parentheses = ('(',')')
        need_product = [parentheses for i in range(2*n)]
        answer = []
        for parentheses in itertools.product(*need_product):
            if self.isPair(parentheses):
                answer.append(''.join(parentheses))
        return answer
